File: src/lib/hours.ts
import { restaurant } from "@/lib/restaurant";

function parseHm(value: string) {
  const [h, m] = value.split(":").map(Number);
  return h * 60 + m;
}

export function getDublinParts(now = new Date()) {
  const parts = new Intl.DateTimeFormat("en-US", {
    timeZone: restaurant.timezone,
    weekday: "long",
    hour: "2-digit",
    minute: "2-digit",
    hourCycle: "h23",
  }).formatToParts(now);

  const weekday = parts.find((p) => p.type === "weekday")?.value ?? "";
  const hour = Number(parts.find((p) => p.type === "hour")?.value ?? "0");
  const minute = Number(parts.find((p) => p.type === "minute")?.value ?? "0");
  return { weekday, minutes: hour * 60 + minute };
}

export function isOpenNow(now = new Date()) {
  const { minutes } = getDublinParts(now);
  const open = parseHm(restaurant.hours.open);
  const close = parseHm(restaurant.hours.close);
  return minutes >= open && minutes < close;
}

export function isToastOrderingOpen(now = new Date()) {
  const { minutes } = getDublinParts(now);
  const open = parseHm(restaurant.hours.open);
  const close = parseHm(restaurant.hours.toastClose);
  return minutes >= open && minutes < close;
}

export function statusCopy(now = new Date()) {
  if (isOpenNow(now)) {
    return {
      open: true,
      label: "Open now",
      detail: `Kitchen until ${restaurant.hours.display.split("–")[1]?.trim() ?? "10:00 PM"}`,
    };
  }
  return {
    open: false,
    label: "Closed now",
    detail: `We open daily at 11:30 AM`,
  };
}

File: raya/hours.py
from datetime import datetime
from zoneinfo import ZoneInfo

from raya.restaurant import restaurant


def _parse_hm(value: str) -> int:
    hours, minutes = value.split(":")
    return int(hours) * 60 + int(minutes)


def dublin_parts(now: datetime | None = None) -> dict:
    now = now or datetime.now(ZoneInfo(restaurant["timezone"]))
    if now.tzinfo is None:
        now = now.replace(tzinfo=ZoneInfo("UTC")).astimezone(ZoneInfo(restaurant["timezone"]))
    else:
        now = now.astimezone(ZoneInfo(restaurant["timezone"]))
    return {
        "weekday": now.strftime("%A"),
        "minutes": now.hour * 60 + now.minute,
    }


def is_open_now(now: datetime | None = None) -> bool:
    minutes = dublin_parts(now)["minutes"]
    return _parse_hm(restaurant["hours"]["open"]) <= minutes < _parse_hm(restaurant["hours"]["close"])


def is_online_ordering_open(now: datetime | None = None) -> bool:
    minutes = dublin_parts(now)["minutes"]
    return _parse_hm(restaurant["hours"]["open"]) <= minutes < _parse_hm(
        restaurant["hours"]["online_close"]
    )


def status_copy(now: datetime | None = None) -> dict:
    if is_open_now(now):
        close_label = restaurant["hours"]["display"].split("–")[-1].strip()
        return {"open": True, "label": "Open now", "detail": f"Kitchen until {close_label}"}
    return {"open": False, "label": "Closed now", "detail": "We open daily at 11:30 AM"}
